cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringProgramEventPublisher.java

```diff
@@ -0,0 +1,232 @@
+  private final MessagingService messagingService;
+  private final MessageFetcher messageFetcher;
+  private final MessagePublisher messagePublisher;
+  private final ProgramRunRecordFetcher runRecordFetcher;
+  private final TransactionRunner transactionRunner;
+  private final String programUpdateTopic;
+  private final int programUpdateFetchSize;
+  private final String programStateTopicPrefix;
+  private final long pollInterval;
+
+  // Tracks id of last program status update message that was processed.
+  private String lastProgramUpdateMessageId;
+
+  @Inject
+  TetheringProgramEventPublisher(CConfiguration cConf, TetheringStore store, MessagingService messagingService,
+                                 ProgramRunRecordFetcher programRunRecordFetcher, TransactionRunner transactionRunner) {
+    super(RetryStrategies.fromConfiguration(cConf, "tethering.agent."));
+    this.store = store;
+    this.messagingService = messagingService;
+    this.messageFetcher = new MultiThreadMessagingContext(messagingService).getMessageFetcher();
+    this.messagePublisher = new MultiThreadMessagingContext(messagingService).getMessagePublisher();
+    this.runRecordFetcher = programRunRecordFetcher;
+    this.transactionRunner = transactionRunner;
+    this.programUpdateTopic = cConf.get(Constants.AppFabric.PROGRAM_STATUS_RECORD_EVENT_TOPIC);
+    this.programUpdateFetchSize = cConf.getInt(Constants.AppFabric.STATUS_EVENT_FETCH_SIZE);
+    this.programStateTopicPrefix = cConf.get(Constants.Tethering.PROGRAM_STATE_TOPIC_PREFIX);
+    // TetheringAgentService reads messages published by TetheringProgramEventPublisher, so use the same
+    // poll interval here
+    this.pollInterval = TimeUnit.SECONDS.toMillis(cConf.getLong(Constants.Tethering.CONNECTION_INTERVAL));
+  }
+
+  @Override
+  protected void doStartUp() {
+    TransactionRunners.run(transactionRunner, context -> {
+      AppMetadataStore appMetadataStore = AppMetadataStore.create(context);
+      lastProgramUpdateMessageId = appMetadataStore.retrieveSubscriberState(programUpdateTopic,
+                                                                            TetheringAgentService.SUBSCRIBER);
+      if (lastProgramUpdateMessageId == null) {
+        // Initialize subscriber based on last message fetched by RuntimeProgramStatusSubscriberService.
+        // This is to avoid fetching existing program history from before TetheringAgent was added
+        String messageId = appMetadataStore.retrieveSubscriberState(programUpdateTopic,
+                                                                    RuntimeProgramStatusSubscriberService.SUBSCRIBER);
+        appMetadataStore.persistSubscriberState(programUpdateTopic, TetheringAgentService.SUBSCRIBER, messageId);
+        lastProgramUpdateMessageId = messageId;
+      }
+    });
+  }
+
+  @Override
+  protected long runTask() throws IOException {
+    List<PeerInfo> peers = store.getPeers().stream()
+      // Ignore peers in REJECTED state.
```

Comment:
what about not_found and pending? shouldn't we only do this for peers in the ACCEPTED state?

Reply:
fixed to only do this for peers in ACCEPTED state